src/org/javarosa/core/model/instance/utils/CsvAsXmlInputStream.java

```diff
@@ -0,0 +1,46 @@
+package org.javarosa.core.model.instance.utils;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayInputStream;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * An adapter that reads a CSV file and provides an XML file as described in
+ * https://github.com/opendatakit/xforms-spec/issues/88.
+ */
+public class CsvAsXmlInputStream extends InputStream {
+    private final InputStream stream;
+
+    public CsvAsXmlInputStream(String path) throws IOException {
+        BufferedReader br = new BufferedReader(new FileReader(path));
+        StringBuilder sb = new StringBuilder();
+        String csvLine = br.readLine();
+
+        sb.append("<root>\n");
+
+        if (csvLine != null) {
+            String[] fieldNames = csvLine.split(",");
+
+            while ((csvLine = br.readLine()) != null) {
+                String[] data = csvLine.split(",");
+                sb.append("\t<item>\n");
+                for (int i = 0; i < fieldNames.length; ++i) {
+                    sb.append(String.format("\t\t<%s>%s</%s>\n", fieldNames[i], data[i], fieldNames[i]));
+                }
+                sb.append("\t</item>\n");
+            }
+        }
+
+        sb.append("</root>\n");
+
+        stream = new ByteArrayInputStream(sb.toString().getBytes(StandardCharsets.UTF_8));
```

Comment:
If we like this approach, we can change  this class so it doesn’t read the whole file into memory.

Reply:
My only concern is that this is reading all bytes from the original input stream to produce an XML document and then, that XML document is parsed (effectively reading the contents twice).

Since each line represents an item, maybe we could manually produce one `TreeElement` for each row which could be added to a root element, and then return it. I think that this could be more memory and time efficient. It could be even faster than the XML loader.

Reply:
This is what I was thinking about: https://github.com/ggalmazor/javarosa/commit/cd7c991f51e2cf5d56264fd004c30554a7b104d2

Reply:
You’re a clever one!